File: packages/ui-compose/ui/src/main/kotlin/com/kinetixui/ui/Timeline.kt
package com.kinetixui.ui

import androidx.compose.foundation.background
import androidx.compose.foundation.layout.Box
import androidx.compose.foundation.layout.Column
import androidx.compose.foundation.layout.Row
import androidx.compose.foundation.layout.heightIn
import androidx.compose.foundation.layout.padding
import androidx.compose.foundation.layout.size
import androidx.compose.foundation.layout.width
import androidx.compose.foundation.shape.CircleShape
import androidx.compose.material3.Text
import androidx.compose.runtime.Composable
import androidx.compose.ui.Alignment
import androidx.compose.ui.Modifier
import androidx.compose.ui.res.dimensionResource
import androidx.compose.ui.text.font.FontWeight
import androidx.compose.ui.unit.dp
import androidx.compose.ui.unit.sp

/**
 * KinetixTimeline — mirrors `packages/ui/src/components/timeline.tsx`:
 * ordered events down a rail (dot, connector, time, content).
 * `alternating` lays content left/right of a centered rail with three
 * equal-weighted columns; the default is a single left-aligned rail. Same
 * "fixed min-height instead of a dynamic stretch" connector simplification
 * as [KinetixStepper]'s vertical orientation — Compose has no cheap
 * equivalent for React's `flex-1 self-stretch` without a custom layout.
 */
data class KinetixTimelineItem(
    val title: String,
    val time: String? = null,
    val content: String? = null,
)

@Composable
fun KinetixTimeline(
    items: List<KinetixTimelineItem>,
    modifier: Modifier = Modifier,
    alternating: Boolean = false,
) {
    Column(modifier = modifier) {
        items.forEachIndexed { i, item ->
            val isLast = i == items.lastIndex
            if (!alternating) {
                Row {
                    TimelineRail(isLast)
                    Column(
                        modifier = Modifier.padding(
                            start = dimensionResource(R.dimen.spacing_3),
                            bottom = if (!isLast) dimensionResource(R.dimen.spacing_6) else 0.dp,
                        ),
                    ) {
                        TimelineBody(item, Alignment.Start)
                    }
                }
            } else {
                val onRight = i % 2 == 0
                Row(verticalAlignment = Alignment.Top) {
                    Box(modifier = Modifier.weight(1f), contentAlignment = Alignment.TopEnd) {
                        if (!onRight) TimelineBody(item, Alignment.End)
                    }
                    TimelineRail(isLast)
                    Box(
                        modifier = Modifier
                            .weight(1f)
                            .padding(start = dimensionResource(R.dimen.spacing_4)),
                        contentAlignment = Alignment.TopStart,
                    ) {
                        if (onRight) TimelineBody(item, Alignment.Start)
                    }
                }
            }
        }
    }
}

@Composable
private fun TimelineRail(isLast: Boolean) {
    val colors = KinetixColorScheme.current
    Column(horizontalAlignment = Alignment.CenterHorizontally) {
        // dot size (10dp) isn't on the shared spacing scale — same call as KinetixBadge.
        Box(modifier = Modifier.size(10.dp).background(colors.action, CircleShape))
        if (!isLast) {
            Box(
                modifier = Modifier
                    .padding(vertical = dimensionResource(R.dimen.spacing_1))
                    .width(dimensionResource(R.dimen.border_width_default))
                    .heightIn(min = dimensionResource(R.dimen.spacing_6)) // fixed floor, not a dynamic stretch — see class doc
                    .background(colors.border),
            )
        }
    }
}

@Composable
private fun TimelineBody(item: KinetixTimelineItem, horizontalAlignment: Alignment.Horizontal) {
    val colors = KinetixColorScheme.current
    Column(horizontalAlignment = horizontalAlignment) {
        if (item.time != null) {
            Text(text = item.time, color = colors.mutedForeground, fontSize = dimensionResource(R.dimen.font_size_label_sm).value.sp)
        }
        Text(
            text = item.title,
            color = colors.foreground,
            fontWeight = FontWeight.Medium,
            fontSize = dimensionResource(R.dimen.font_size_label_md).value.sp,
        )
        if (item.content != null) {
            Text(text = item.content, color = colors.mutedForeground, fontSize = dimensionResource(R.dimen.font_size_body_sm).value.sp)
        }
    }
}

File: packages/ui-swiftui/Sources/KinetixUI/Timeline.swift
//
// Timeline.swift — KinetixTimeline.
//
// Mirrors packages/ui/src/components/timeline.tsx: ordered events down a
// rail (dot, connector, time, content). `alternating` lays content
// left/right of a centered rail; the default is a single left-aligned
// rail. Same fixed-min-height connector call as KinetixStepper's
// vertical orientation — SwiftUI's Rectangle has no cheap "stretch to
// fill the row" without a custom layout either.
//

import SwiftUI

public struct KinetixTimelineItem {
    public let title: String
    public let time: String?
    public let content: String?

    public init(title: String, time: String? = nil, content: String? = nil) {
        self.title = title
        self.time = time
        self.content = content
    }
}

public struct KinetixTimeline: View {
    @Environment(\.kinetixColors) private var colors

    private let items: [KinetixTimelineItem]
    private let alternating: Bool

    public init(items: [KinetixTimelineItem], alternating: Bool = false) {
        self.items = items
        self.alternating = alternating
    }

    public var body: some View {
        let indexed = Array(items.enumerated())
        VStack(alignment: .leading, spacing: 0) {
            ForEach(indexed, id: \.offset) { idx, item in
                if alternating {
                    alternatingRow(idx: idx, item: item, isLast: idx == items.count - 1)
                } else {
                    row(idx: idx, item: item, isLast: idx == items.count - 1)
                }
            }
        }
    }

    @ViewBuilder
    private func rail(isLast: Bool) -> some View {
        VStack(spacing: 4) {
            Circle().fill(colors.action).frame(width: 10, height: 10)
            if !isLast {
                Rectangle()
                    .fill(colors.border)
                    .frame(width: 1)
                    .frame(minHeight: 24) // fixed floor, not a dynamic stretch — see file doc
            }
        }
    }

    @ViewBuilder
    private func body(_ item: KinetixTimelineItem, alignment: HorizontalAlignment) -> some View {
        VStack(alignment: alignment, spacing: 2) {
            if let time = item.time {
                Text(time).font(.kinetixLabelSm).foregroundStyle(colors.mutedForeground)
            }
            Text(item.title).font(.kinetixLabelMd.weight(.medium)).foregroundStyle(colors.foreground)
            if let content = item.content {
                Text(content).font(.kinetixBodySm).foregroundStyle(colors.mutedForeground)
            }
        }
    }

    @ViewBuilder
    private func row(idx: Int, item: KinetixTimelineItem, isLast: Bool) -> some View {
        HStack(alignment: .top, spacing: 12) {
            rail(isLast: isLast)
            body(item, alignment: .leading)
                .padding(.bottom, isLast ? 0 : 24)
        }
    }

    @ViewBuilder
    private func alternatingRow(idx: Int, item: KinetixTimelineItem, isLast: Bool) -> some View {
        let onRight = idx % 2 == 0
        HStack(alignment: .top, spacing: 16) {
            if onRight {
                Color.clear.frame(maxWidth: .infinity)
            } else {
                body(item, alignment: .trailing).frame(maxWidth: .infinity, alignment: .trailing)
            }
            rail(isLast: isLast)
            if onRight {
                body(item, alignment: .leading).frame(maxWidth: .infinity, alignment: .leading)
            } else {
                Color.clear.frame(maxWidth: .infinity)
            }
        }
    }
}
